plonk-core/src/constraint_system/composer.rs

```diff
@@ -526,6 +526,74 @@ where
         self.n += 1;
     }
 
+    /// This function is used to add a blinding factors to the witness
+    /// and permutation polynomials.
+    /// All gate selectors are turned off to guarantee the constraints
+    /// are still satisfied.
+    pub fn add_blinding_factors(&mut self) {
```

Comment:
Can this take the RNG as a parameter?

```rust
pub fn add_blinding_factors<R>(&mut self, rng: &mut R)
where
    R: CryptoRng + RngCore + ?Sized;
```

Reply:
Also, since the gate selectors are off, we have to make sure this doesn't get disabled by the "empty selectors shouldn't contribute to the proof" mechanism. Perhaps we need to check that there is a non-trivial permutation on the specific gate as well.

Reply:
I don't think that the unused selectors optimizations will interfere with the blinders.
The way I see it, those optimizations only affect:
 -  the quotient and lin polys, which are not affected by the blinders anyways 
 -  the proof and verifier keys which may omit some evaluations/commitments of unnecessary selectors, but everything witness realted stays unchanged.
 
 In regards to the non-trivial permutation I'm still working on guaranteeing that we wont get the trivial permutation in the whole circuit. (An attempt prevent this is the addition of the `zero_var()` in the 3rd blinder gate: https://github.com/ZK-Garage/plonk/pull/118/files#diff-7580c65a30bf9cd380a7b494e25c0fd4b5760734af32c97a485d9e1d9b52b1f8R592-R598).
 Is this what you mean by "non-trivial permutation on the specific gate"?